### PR TITLE
goto instrumentation for decreases clauses on loops

### PR DESCRIPTION
**Pull request by LongPham7**
This pull request contains the full-fledged implementation of one-dimensional loop variants (also known as ranking functions). A loop variant is an integer expression (i) that is bounded below and (ii) strictly decreases after each iteration of a loop. We use loop variants to prove termination of loops. 

A loop variant in C is converted to the following statements in the GOTO language: 
1. Declarations of two temporary variables
2. Assign a loop variant to the first temporary variable at the "beginning" of a loop body
3. Assign a loop variant to the second temporary variable at the "end" of a loop body
4. Assert that the second temporary variable < first temporary variable
5. Discard the two temporary variables by the "dead" statements in the GOTO language. 

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-programs/goto_program.h [hunk 1]
```diff
@@ -1078,6 +1079,17 @@ class goto_programt
       nil_exprt(),
       {});
   }
+
+  // Returns an assertion smaller < larger. 
+  // This is necessary for the implementation of loop variants. 
+  static instructiont make_inequality(
```
Comment (SaswatPadhi) [inline src/goto-programs/goto_program.h]: This function is only ever called once, so why don't we just inline its definition there? It's a single line anyway ...

File: src/goto-programs/goto_convert.cpp [hunk 2]
```diff
@@ -857,25 +857,40 @@ void goto_convertt::convert_assume(
   dest.add(goto_programt::make_assumption(op, code.source_location()));
 }
 
-void goto_convertt::convert_loop_invariant(
+void goto_convertt::convert_loop_invariant_and_decreases(
```
Comment (SaswatPadhi) [inline src/goto-programs/goto_convert.cpp]: I'd just call it `convert_loop_contracts` now. Assigns clause stuff would also be added in this function eventually

File: src/goto-programs/goto_convert.cpp [hunk 3]
```diff
@@ -857,25 +857,40 @@ void goto_convertt::convert_assume(
...
+    PRECONDITION(loop->is_goto());
+    loop->guard.add(ID_C_spec_loop_invariant).swap(invariant);
   }
 
-  PRECONDITION(loop->is_goto());
-  loop->guard.add(ID_C_spec_loop_invariant).swap(invariant);
+  if(!variant.is_nil())
+  {
+    // Interestingly enough, a function call is considered a side effect even if
+    // the callee entails no side effects.
```
Comment (SaswatPadhi) [inline src/goto-programs/goto_convert.cpp]: I'd remove these two  lines -- this is a known issue. If you want to document this, let's document that before the first call to `has_subexpr` as well. May be add a `TODO` at line 868 and explain this, if you want to.

File: src/goto-instrument/code_contracts.h [hunk 4]
```diff
@@ -98,12 +98,14 @@ class code_contractst
     const irep_idt &function_id,
     const irep_idt &mode);
 
-  void check_apply_invariants(
+  void check_apply_invariants_and_decreases(
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.h]: I'd just call it `check_apply_loop_contracts` now. Assigns clause stuff would also be added in this function eventually

File: src/goto-instrument/code_contracts.cpp [hunk 5]
```diff
@@ -129,11 +130,18 @@ void code_contractst::check_apply_invariants(
...
-  if(invariant.is_nil())
-    return;
+  exprt variant = static_cast<const exprt &>(
+    loop_end->get_condition().find(ID_C_spec_decreases));
+  if(invariant.is_nil()) 
+  {
+    if (variant.is_nil())
+      return;
+    else 
+      invariant = true_exprt(); // The Boolean true expression
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: [nit]

remove comment; self-explanatory

File: src/goto-instrument/code_contracts.cpp [hunk 6]
```diff
@@ -129,11 +130,18 @@ void code_contractst::check_apply_invariants(
...
     loop_end->get_condition().find(ID_C_spec_loop_invariant));
-  if(invariant.is_nil())
-    return;
+  exprt variant = static_cast<const exprt &>(
+    loop_end->get_condition().find(ID_C_spec_decreases));
+  if(invariant.is_nil()) 
+  {
+    if (variant.is_nil())
+      return;
+    else
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: I'd also log a warning here that we are using a default `true` invariant . It's not unsound -- if insufficient this would fail the checks; but just might be surprising to users.

File: src/goto-programs/goto_program.cpp [hunk 7]
```diff
@@ -117,9 +117,17 @@ std::ostream &goto_programt::output_instruction(
...
     break;
+  case ASSIGN:
+    out << from_expr(ns, identifier, instruction.get_code());
+    {
+      const irep_idt &commentw = instruction.source_location.get_comment();
+      if (!commentw.empty())
+        out << " // " << commentw;
+    }
+    out << '\n'; 
+    break;
```
Comment (SaswatPadhi) [inline src/goto-programs/goto_program.cpp]: Why was this change needed?

File: src/goto-instrument/code_contracts.cpp [hunk 8]
```diff
@@ -209,11 +241,35 @@ void code_contractst::check_apply_invariants(
...
+    havoc_code.add(goto_programt::make_assignment(new_variant_symbol, variant, declaration_new_variant_source_location));
+    havoc_code.instructions.back().source_location.set_comment("Evaluate and store the loop variant after the loop");
+
+    // Generate: assertion that the loop variant after the loop is smaller than
+    // the loop variant before the loop
+    source_locationt assignment_source_location {loop_head->source_location}; 
+    havoc_code.add(goto_programt::make_inequality(new_variant_symbol, old_variant_symbol, assignment_source_location));
+    havoc_code.instructions.back().source_location.set_comment("Compare the old and new variant values");
+
+    // Discartd the temporary variables that store loop variants
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]:
```suggestion
    // Discard the temporary variables that store loop variant values
```

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
Some initial comments below, and could you please:
- fix the formatting issues? (see the build failures for details)
- change the message `Compare the old and new variant values` to `Check that loop variant monotonically decreases`?

File: src/goto-programs/goto_program.cpp (hunk 7 above)
Comment (LongPham7) [reply src/goto-programs/goto_program.cpp]: When we display a GOTO program by the command `goto-instrument --show-goto-functions`, I want to show a comment `// Evaluate and store the loop variant before the loop` next to the assignment for the first temporary variable, and likewise a comment `// Evaluate and store the loop variant after the loop` next to the second temporary variable's assignment. This is why I modified the function `output_instruction` in `goto_program.cpp` such that it can print out not only comments of assertions but also comments of assignments. If you think this feature is redundant, please let me know!

File: src/goto-programs/goto_program.h (hunk 1 above)
Comment (LongPham7) [reply src/goto-programs/goto_program.h]: Good point. Yes, I will fix it.

File: src/goto-programs/goto_convert.cpp (hunk 3 above)
Comment (LongPham7) [reply src/goto-programs/goto_convert.cpp]: OK. I will just remove these two lines of comments.

File: src/goto-instrument/code_contracts.cpp (hunk 6 above)
Comment (LongPham7) [reply src/goto-instrument/code_contracts.cpp]: Yes, I will do it.

File: src/goto-instrument/code_contracts.cpp [hunk 9]
```diff
@@ -129,11 +130,24 @@ void code_contractst::check_apply_invariants(
...
+  {
+    if(variant.is_nil())
+      return;
+    else
+    {
+      invariant = true_exprt();
+      log.warning() << "A loop in function '" << function_name
+                    << "' does not have a loop invariant, but a loop variant. "
+                    << "Hence, a loop invariant is set to true by default. "
+                    << messaget::eom;
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: How about:

```c
      log.warning() << "The loop at " << loop_head->source_location.as_string()
                    << "does not have a loop invariant, but has a loop variant. "
                    << "Hence, a default loop invariant ('true') is being used."
                    << messaget::eom;
```

File: src/goto-instrument/code_contracts.cpp [hunk 10]
```diff
@@ -129,11 +130,24 @@ void code_contractst::check_apply_invariants(
...
+    loop_end->get_condition().find(ID_C_spec_decreases));
   if(invariant.is_nil())
-    return;
+  {
+    if(variant.is_nil())
+      return;
+    else
+    {
+      invariant = true_exprt();
+      log.warning() << "A loop in function '" << function_name
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: Could you please also add a regression test to check if this warning is correctly triggered?

File: src/goto-programs/goto_program.cpp (hunk 7 above)
Comment (SaswatPadhi) [reply src/goto-programs/goto_program.cpp]: We usually don't add comments on assignments, but only on the assertions (which you're doing already in code_contracts.cpp). The comments on assignments are not guaranteed to be preserved across subsequent transformations (simplification passes etc.), so may be we can remove these comments and revert this printing logic.

File: src/goto-instrument/code_contracts.cpp [hunk 11]
```diff
@@ -188,6 +205,37 @@ void code_contractst::check_apply_invariants(
     converter.goto_convert(assumption, havoc_code, mode);
   }
 
+  // Temproary variables that store the variant before and after the loop
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]:
```suggestion
  // Temporary variables that store the variant before and after the loop
```

File: src/goto-instrument/code_contracts.cpp [hunk 12]
```diff
@@ -188,6 +205,37 @@ void code_contractst::check_apply_invariants(
...
+    // variant before the loop entry
+    havoc_code.add(
+      goto_programt::make_decl(old_variant_symbol, loop_head->source_location));
+
+    // Generate: an assignment to store the loop variant before the loop entry
+    // We would like to set a comment associated with a source location.
+    // However, we have already used loop_head->source_location to set a
+    // comment. Furthermore, the third argument of make_assignment is passed by
+    // reference rather than by value. Therefore, it is necessary to clone
+    // loop_head->source_location.
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: I don't understand the issue here. What exactly is the problem with passing the `source_location` by value? Is it that we are adding a comment? If so, please remove the comment on the assignment (I explained the rationale in another comment below).

(same comment for the other assignment below)

File: src/goto-instrument/code_contracts.cpp [hunk 13]
```diff
@@ -188,6 +205,37 @@ void code_contractst::check_apply_invariants(
...
+    source_locationt declaration_old_variant_source_location{
+      loop_head->source_location};
+    havoc_code.add(goto_programt::make_assignment(
+      old_variant_symbol, variant, declaration_old_variant_source_location));
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: Instead of generating a goto instruction directly, we should be calling `goto_convert` to properly translate a high-level assignment to appropriate goto instructions. Use:

```c++
    code_assignt old_variant_assignment{old_variant_symbol, variant};
    old_variant_assignment.add_source_location() = loop_head->source_location;
    converter.goto_convert(old_variant_assignment, havoc_code, mode);
```

(same for the other assignment)

File: src/goto-instrument/code_contracts.cpp [hunk 14]
```diff
@@ -209,14 +257,47 @@ void code_contractst::check_apply_invariants(
...
+      new_variant_symbol, variant, declaration_new_variant_source_location));
+    havoc_code.instructions.back().source_location.set_comment(
+      "Evaluate and store the loop variant after the loop");
+
+    // Generate: assertion that the loop variant after the loop is smaller than
+    // the loop variant before the loop
+    source_locationt assignment_source_location{loop_head->source_location};
+    havoc_code.add(goto_programt::make_assertion(
+      binary_relation_exprt(new_variant_symbol, ID_lt, old_variant_symbol),
+      assignment_source_location));
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: Instead of generating a goto instruction directly, we should be calling `goto_convert` to properly translate a high-level assertion to appropriate goto instructions. Use:

```c
    code_assertt variant_assertion{binary_relation_exprt(new_variant_symbol, ID_lt, old_variant_symbol)};
    variant_assertion.add_source_location() = loop_end->source_location;
    converter.goto_convert(variant_assertion, havoc_code, mode);
```

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]

File: src/goto-programs/goto_program.cpp (hunk 7 above)
Comment (LongPham7) [reply src/goto-programs/goto_program.cpp]: OK. I will revert the change I had made.

File: src/goto-instrument/code_contracts.cpp (hunk 10 above)
Comment (LongPham7) [reply src/goto-instrument/code_contracts.cpp]: Yes, I will.

File: src/goto-instrument/code_contracts.cpp (hunk 14 above)
Comment (LongPham7) [reply src/goto-instrument/code_contracts.cpp]: Thanks a lot for the code.

File: src/goto-instrument/code_contracts.cpp (hunk 12 above)
Comment (LongPham7) [reply src/goto-instrument/code_contracts.cpp]: Now that we use `goto_convert` instead of `make_assignment`, we no longer need to clone a `source_location` object. So this issue is now gone.

File: src/goto-instrument/code_contracts.cpp [hunk 15]
```diff
@@ -129,11 +130,25 @@ void code_contractst::check_apply_invariants(
...
-    return;
+  {
+    if(variant.is_nil())
+      return;
+    else
+    {
+      invariant = true_exprt();
+      log.warning()
+        << "The loop at " << loop_head->source_location.as_string()
+        << " does not have a loop invariant, but has a loop variant. "
```
Comment (martin-cs) [inline src/goto-instrument/code_contracts.cpp]: Please can you be consistent with naming?  Personally I don't like "loop variant" but I care most about it being consistent.  SO if you are going to call it a loop variant in the help text and comments then maybe it should be `__CPROVER_loop_variant`.

File: src/goto-instrument/code_contracts.cpp [hunk 16]
```diff
@@ -188,6 +206,29 @@ void code_contractst::check_apply_invariants(
...
+  {
+    // Generate: a declaration of the temporary variable that stores the loop
+    // variant before the loop entry
+    havoc_code.add(
+      goto_programt::make_decl(old_variant_symbol, loop_head->source_location));
+
+    // Generate: an assignment to store the loop variant before the loop entry
+    code_assignt old_variant_assignment{old_variant_symbol, variant};
+    old_variant_assignment.add_source_location() = loop_head->source_location;
+    converter.goto_convert(old_variant_assignment, havoc_code, mode);
```
Comment (martin-cs) [inline src/goto-instrument/code_contracts.cpp]: @tautschnig is this use of `goto_convert` fitting with how and when / where it is supposed to be used?

**Review by martin-cs — APPROVED** [review]
Looks good apart from the things flagged.

File: src/goto-instrument/code_contracts.cpp (hunk 15 above)
Comment (LongPham7) [reply src/goto-instrument/code_contracts.cpp]: Thanks for your suggestion. As of now, the concrete syntax for loop variants is `__CPROVER_decreases(...)`. However, "decreases" is not a noun (or at least, it would be awkward to use it as a noun). This is why I currently use the phrase "loop variants" in the code/comments to refer to the loop variants/ranking functions. I will discuss this with Saswat.

Comment (LongPham7) [reply src/goto-instrument/code_contracts.cpp]: @martin-cs I misunderstood your comment - I thought you were suggesting changing all function names and comments in the code base so that we never mention "loop variants", but now I believe you are just referring to the warning issued to a user. Yes, I will change the warning message to something like "The loop at ... does not have a loop invariant, but has a decreases-clause." In the code base, however, I will keep using the phrase "loop variants" in comments and function names.

Comment (martin-cs) [reply src/goto-instrument/code_contracts.cpp]: So, I am very bad at using consistent terminology and it causes confusion.  My comments, here and earlier, are intended to try to help you avoid a similar fate.

I would /suggest/ that you use one terminology for *everything*; the annotation, the user error reporting, in the comments, in the code, everywhere.  Either call them ranking functions, or loop variants, or ... whatever name you want.  Personally I feel that loop variant can cause confusion with loop invariants but that is just my view.  I think it is far more important to be consistent than it is to pick one particular name.

Comment (LongPham7) [reply src/goto-instrument/code_contracts.cpp]: OK. Thanks for the clarification! I will update my comments and function names in the code.

Comment (LongPham7) [reply src/goto-instrument/code_contracts.cpp]: I've just updated all function names, comments, and warning messages in the code such that the phrase "DECRAESES clause" is used in place of "loop variants."

File: regression/contracts/invar_check_multiple_loops/test.desc [hunk 17]
```diff
@@ -5,8 +5,10 @@ main.c
 ^SIGNAL=0$
 ^\[main.1\] .* Check loop invariant before entry: SUCCESS$
 ^\[main.2\] .* Check that loop invariant is preserved: SUCCESS$
-^\[main.3\] .* Check loop invariant before entry: SUCCESS$
-^\[main.4\] .* Check that loop invariant is preserved: SUCCESS$
+^\[main.3\] .* Check that DECREASES clause monotonically decreases: SUCCESS$
```
Comment (SaswatPadhi) [inline regression/contracts/invar_check_multiple_loops/test.desc]: We don't capitalize the other contracts to this looks a bit odd. How about:

```
Check monotonicity of decreases clause: ...
```

Comment (LongPham7) [reply regression/contracts/invar_check_multiple_loops/test.desc]: Yeah, initially, I didn't capitalize "decreases clauses". However, in that warning message, we use "decreases" twice, and I thought it might confuse the user. This is why I decided to capitalize one of the two occurrences of "decreases." Your suggested warning message is much better. I will modify the code accordingly.

File: src/goto-programs/goto_program.cpp [hunk 18]
```diff
@@ -131,7 +131,7 @@ std::ostream &goto_programt::output_instruction(
     {
       out << from_expr(ns, identifier, instruction.get_condition());
 
-      const irep_idt &comment=instruction.source_location.get_comment();
+      const irep_idt &comment = instruction.source_location.get_comment();
```
Comment (SaswatPadhi) [inline src/goto-programs/goto_program.cpp]: [nit]

Please undo this change.

File: src/goto-programs/goto_convert_class.h [hunk 19]
```diff
@@ -236,7 +236,7 @@ class goto_convertt:public messaget
     goto_programt &dest,
     const irep_idt &mode);
   void convert_cpp_delete(const codet &code, goto_programt &dest);
-  void convert_loop_invariant(
+  void convert_loop_contract(
```
Comment (SaswatPadhi) [inline src/goto-programs/goto_convert_class.h]:
```suggestion
  void convert_loop_contracts(
```

(here and elsewhere)

File: src/goto-programs/goto_convert.cpp [hunk 20]
```diff
@@ -857,25 +857,40 @@ void goto_convertt::convert_assume(
...
   }
 
-  PRECONDITION(loop->is_goto());
-  loop->guard.add(ID_C_spec_loop_invariant).swap(invariant);
+  if(!decreases_clause.is_nil())
+  {
+    if(has_subexpr(decreases_clause, ID_side_effect))
+    {
+      throw incorrect_goto_program_exceptiont(
+        "decreases clause is not side-effect free.",
```
Comment (SaswatPadhi) [inline src/goto-programs/goto_convert.cpp]:
```suggestion
        "Decreases clause is not side-effect free.",
```

**Review by SaswatPadhi — APPROVED** [review]
Couple of minor comments, but otherwise LGTM! Thanks for all the work!

**Review by feliperodri — APPROVED** [review]
Thank you for tackling all comments!

File: src/goto-instrument/code_contracts.cpp [hunk 21]
```diff
@@ -209,14 +253,46 @@ void code_contractst::check_apply_invariants(
...
+
+    // Generate: assertion that the decreases clause's value after the loop is
+    // smaller than the value before the loop
+    code_assertt monotonic_decreasing_assertion{
+      binary_relation_exprt(new_decreases_symbol, ID_lt, old_decreases_symbol)};
+    monotonic_decreasing_assertion.add_source_location() =
+      loop_head->source_location;
+    converter.goto_convert(monotonic_decreasing_assertion, havoc_code, mode);
+    havoc_code.instructions.back().source_location.set_comment(
+      "Check monotonicity of decreases clause");
```
Comment (kroening) [inline src/goto-instrument/code_contracts.cpp]: "monotonicity" is perhaps not the right word. It might not be an intuitive description for a broader audience, and being pedantic, you need strict monotonicity to get termination.
How about "ranking function decreases"?

**Comment by kroening** [comment]
What should happen when there are multiple `__CPROVER_decreases` clauses for one loop?
You could make that a lexicographic ranking function.

**Review by kroening — APPROVED** [review]

**Comment by SaswatPadhi** [comment]
> What should happen when there are multiple `__CPROVER_decreases` clauses for one loop?
You could make that a lexicographic ranking function.

Yes, this is a good point. Currently we don't allow multiple invariant and decreases annotations, but my proposal was also to treat multiple decreases clause annotations as tuples that would be compared lexicographically.

File: src/goto-instrument/code_contracts.cpp (hunk 21 above)
Comment (SaswatPadhi) [reply src/goto-instrument/code_contracts.cpp]: I think "ranking function" might also sound unfamiliar, plus we haven't used ranking function elsewhere (the annotation just reads `__CPROVER_decreases`).

How about simply, `"Check decreases clause on loop iteration"`?

Comment (LongPham7) [reply src/goto-instrument/code_contracts.cpp]: Thanks a lot for both of your suggestions. I will change the comment to `"Check decreases clause on loop iteration"`.
